### ISSUE TITLE
[symfony/cache] Big time shifting for items with small TTL

### ISSUE THREAD

**Issue by ilyachase** [symfony/symfony#31573]
**Description**  
Recently we've faced problem with memcached expiry time shifting by +-1 second. This problem corresponds [with fact that there is no resolution clock in memcached](https://github.com/memcached/memcached/issues/307). So if I set cache with expire time 3 seconds, it will be there at least 2. But when I use symfony/cache wrapper, I as a rule get result with less than 2 seconds. And I don't know why yet. One of the reason may be long execution stack, another is that in Memcached there are two ways to set expiration time: in seconds less than seconds in 30 days, and in timestamp. In symfony/cache wrapper, expiry always sets as time() + expiry_seconds (second option) here: `symfony/cache/CacheItem.php:97`.

**How to reproduce**  
put following code inside test.php:
```php
<?php

require_once 'vendor/autoload.php';

use Symfony\Component\Cache\Adapter\MemcachedAdapter;

$m = new Memcached();
$m->addServer('localhost', 11211);

$m->set('test', true, (int)$argv[1]);

$time = 0;
while ($m->get('test')) {
    usleep(100 * 1000);
    $time += 0.1;
    var_dump("Memcached: $time");
}

$cacheClient = MemcachedAdapter::createConnection(['memcached://localhost:11211']);
$cache = new MemcachedAdapter($cacheClient);

$item = $cache->getItem('test');
$item->expiresAfter((int)$argv[1]);
$cache->save($item);

$time = 0;
while ($cache->getItem('test')->isHit()) {
    usleep(100 * 1000);
    $time += 0.1;
    var_dump("Symfony: $time");
}
```

`docker run -p 11211:11211 -d memcached`
`docker run -it --rm --network=host -v $PWD:/src -w /src chialab/php:7.3 php test.php 3`

**Example output**
```
string(14) "Memcached: 0.1"                                                                                                                                             
string(14) "Memcached: 0.2"                                                                                                                                             
string(14) "Memcached: 0.3"                                                                                                                                             
string(14) "Memcached: 0.4"                                                                                                                                             
...                                                                                                                                    
string(14) "Memcached: 2.6"                                                                                                                                             
string(14) "Memcached: 2.7"                                                                                                                                             
string(14) "Memcached: 2.8"                                                                                                                                             
string(12) "Symfony: 0.1"                                                                                                                                               
string(12) "Symfony: 0.2"                                                                                                                                               
string(12) "Symfony: 0.3"                                                                                                                                               
...                                                                                                                                    
string(12) "Symfony: 1.7"                                                                                                                                               
string(12) "Symfony: 1.8"
```

You can run it hundred times and through symfony/cache, cache will not leave at least 2 seconds.

**Comment by Tobion** [comment]
According to https://github.com/memcached/memcached/issues/307#issuecomment-341415768 it's +0/-0.999999 and the rest of the lost time is probably network and execution overhead. What is your expected result? To get closer to 3 seconds ttl? What implications does this have for you?

**Comment by ilyachase** [comment]
It's not network overheard because I tried the same with Doctrine cache and pure Memcached. It's probably execution overhead of symfony/cache. My expectations are that such thin shim should not affect that much.

**Comment by ilyachase** [comment]
Also It might be not just overhead but the way how symfony/cache sets up `expire` (Memcached supports two ways buy symfony/cache uses only one).

**Comment by nicolas-grekas** [comment]
The description is correct: the expiry is 1 second less than specified. This is legal of course, and it prevents any issues with going past the lifetime, because of second resolution of the clock of the backend.
Not sure we can fix this, but if you manage to, PR welcome.

**Comment by ilyachase** [comment]
@nicolas-grekas did you read initial issue carefully? I said with symfony/cache 1 seconds become 2 seconds.

**Comment by nicolas-grekas** [comment]
I read the opposite: you set 3s but the item lasts less than 2. What do I miss?

**Comment by ilyachase** [comment]
I mean negotive time shift became 2 seconds.
But I still don't unserstaand - so symfony/cache adds this +1 second shift on purpose? How can you go past the lifetime if memcached only can shift netogive (+0/-0.999999), not positive?

**Comment by nicolas-grekas** [comment]
You can't, but Memcached is a backend that does it correctly. The +1s shift is in a generic code that doesn't know if backends are properly implemented. So it takes the safe side and assumes not.
I suppose this can be updated, please have a look if you want.

**Comment by ilyachase** [comment]
Thanks for explonation, I will take a look.

**Comment by Simperfit** [comment]
@ilyachase did you have a look ?

**Comment by ilyachase** [comment]
@Simperfit yeah. So this is happening as following:
1. You set expiration using `expiresAfter(3)`. Since `3` is int, at line `CacheItem.php:103` expiry property will be set to some float number, let's say `1569609750.766`.
2. Then, when items being saved, at `Adapter/AbstractAdapter.php:147` it will call closure defined at `Adapter/AbstractAdapter.php:69`. In this closure at `Adapter/AbstractAdapter.php:79` $ttl defines as `(int) ($item->expiry - $now)`. $now is another result of `microtime(true)` but it was called *a little bit later* after you called `expiresAfter(3)`. So in $now variable will be `a little bit bigger` float number than was used when calculating expiry at step 1. And then it casts result to int.
In the dump it looks like:
```
1569609750.766 // $item->expiry
1569609747.766 // $now
2.9999709129333 // subtraction result
2 // subtraction result casted by int
```
Therefore it is a symfony/cache logic subtract one more second from items lifetime.

Not sure how to fix it. One of the solutions is to call `time()` instead of `microtime(true)` at `Adapter/AbstractAdapter.php:72`. It will always get seconds count rounded to lesser side, so after subtraction and casting to int you will get more "normal" result without subtraction one more second. In my tests it always gives `3.x` seconds after subtraction and always casts to 3 after `(int)` using. May be @nicolas-grekas have better ideas.

**Comment by nicolas-grekas** [comment]
Fixed in #33846
